main.go

```diff
@@ -1,4 +1,278 @@
+	}
+
+	fmt.Fprintf(feedback.Out(), "Uploading certificates to %s...\n", portAddress)
+
+	certificateSketchFile, err := paths.WriteToTempFile(certificateSketch, paths.TempDir(), "portenta-c33-fwuploader-plugin")
+	if err != nil {
+		return err
+	}
+	defer certificateSketchFile.Remove()
+
+	if err := d.uploadSketch(&portAddress, feedback, certificateSketchFile); err != nil {
+		return err
+	}
+
+	// Open connection
+	connection, err := serialx.Open(portAddress, &serialx.Mode{BaudRate: 115200})
+	if err != nil {
+		return err
+	}
+	if err := connection.SetReadTimeout(100 * time.Second); err != nil {
+		return err
+	}
+
+	time.Sleep(1 * time.Second)
+
+	// Send initial message
+	if _, err := connection.Write([]byte("Y\r\n")); err != nil {
+		return err
+	}
+
+	// reads for possible errors coming from the sketch
+	{
+		var resultSerialOutput string
+		buffer := make([]byte, 64)
+		for {
+			n, err := connection.Read(buffer)
+			if err != nil {
+				return err
+			}
+			if n == 0 {
+				break
+			}
+			resultSerialOutput += string(buffer[0:n])
+			if strings.Contains(resultSerialOutput, "YSTART") {
+				break
+			}
+		}
+
+		if strings.Contains(resultSerialOutput, "ERR:") {
+			errs := strings.Split(resultSerialOutput, "\r\n")
+			result := fmt.Sprintf("%v", errs[1][4:])
+			// the ignore the last message as it's not an error
+			for _, e := range errs[1 : len(errs)-1] {
+				result += ", " + e[4:]
+			}
+			return errors.New(result)
+		}
+	}
+
+	data, err := certificatePath.ReadFile()
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(feedback.Out(), "Please wait a few seconds...\n")
+
+	time.Sleep(1 * time.Second)
+
+	if err := ymodem.ModemSend(connection, data, certificatePath.Base()); err != nil {
+		return err
+	}
+
+	fmt.Fprintf(feedback.Out(), "Upload completed!\n")
+	return nil
+}
+
+// GetFirmwareVersion retrieve the firmware version installed on the board
+func (d *portentaC33Plugin) GetFirmwareVersion(portAddress, fqbn string, feedback *helper.PluginFeedback) (*semver.RelaxedVersion, error) {
+	if portAddress == "" {
+		fmt.Fprintln(feedback.Err(), "Port address not specified")
+		return nil, fmt.Errorf("invalid port address")
+	}
+	fmt.Fprintf(feedback.Out(), "Getting firmware version from %s...\n", portAddress)
+
+	versionSketchFile, err := paths.WriteToTempFile(versionSketch, paths.TempDir(), "portenta-c33-fwuploader-plugin")
+	if err != nil {
+		return nil, err
+	}
+	defer versionSketchFile.Remove()
+
+	if err := d.uploadSketch(&portAddress, feedback, versionSketchFile); err != nil {
+		return nil, err
+	}
+
+	port, err := serialx.Open(portAddress, &serialx.Mode{
+		BaudRate: 9600,
+		Parity:   serialx.NoParity,
+		DataBits: 8,
+		StopBits: serialx.OneStopBit,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer port.Close()
+
+	var version string
+	scanner := bufio.NewScanner(port)
+	for scanner.Scan() {
+		version = scanner.Text()
+		break
+	}
```

Comment:
Uhm... this loop may block forever if the sketch crashes, also the serial port does not behave nicely with `bufio.Scanner`.

I'd change this with just:
```go
// wait 1 second to allow the sketch to fill the buffer with the version string
time.Sleep(time.Second)

buff := make([]byte, 30)
port.SetReadTimeout(time.Second)
n, err := port.Read(buff)
if err != nil || n==0 { .... }
version = strings.TrimSpace(string(buff))
```